src/libsyntax/parse/parser.rs

```diff
@@ -6666,6 +6735,34 @@ impl<'a> Parser<'a> {
                                     maybe_append(attrs, extra_attrs));
             return Ok(Some(item));
         }
+
+        // `unsafe async fn` or `async fn`
+        if (
+            self.check_keyword(keywords::Unsafe) &&
+            self.look_ahead(1, |t| t.is_keyword(keywords::Async))
```

Comment:
Looks like edition checks are missing here and below.

Reply:
That's intentional-- `async fn` is supposed to work on edition 2015. It's just closures and blocks that don't.